File: DAY 3/solutions.py/918. Maximum Sum Circular Subarray.py
class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:

        total = sum(nums)

        curMax = maxSum = nums[0]
        curMin = minSum = nums[0]

        for i in range(1, len(nums)):

            curMax = max(nums[i], curMax + nums[i])
            maxSum = max(maxSum, curMax)

            curMin = min(nums[i], curMin + nums[i])
            minSum = min(minSum, curMin)

        if maxSum < 0:
            return maxSum

        return max(maxSum, total - minSum)

File: DAY 3/solutions.java/918. Maximum Sum Circular Subarray.java
class Solution {
    public int maxSubarraySumCircular(int[] nums) {

        int total = 0;

        int curMax = nums[0];
        int maxSum = nums[0];

        int curMin = nums[0];
        int minSum = nums[0];

        for (int num : nums) {

            total += num;
        }

        for (int i = 1; i < nums.length; i++) {

            curMax = Math.max(nums[i], curMax + nums[i]);
            maxSum = Math.max(maxSum, curMax);

            curMin = Math.min(nums[i], curMin + nums[i]);
            minSum = Math.min(minSum, curMin);
        }

        if (maxSum < 0) {
            return maxSum;
        }

        return Math.max(maxSum, total - minSum);
    }
}